In [1]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf

from scipy import stats

In [2]:
# Carregando a base de gorjetas
tips = sns.load_dataset("tips")

#Criando a variável tip_pct - % da Gorjeta em Relação ao Valor Líquido da Conta
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])

# Criando a variável net_bill - Valor Líquido da Conta
tips['net_bill'] = tips['total_bill'] - tips['tip']


# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [3]:
# Reproduzindo a Regressão de tip Explicada por net_bill
regression_model = smf.ols(formula='tip ~ net_bill', data=tips).fit()
print(regression_model.summary())

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.330
Method:                 Least Squares   F-statistic:                     120.6
Date:                Sat, 22 Jul 2023   Prob (F-statistic):           5.02e-23
Time:                        20:19:24   Log-Likelihood:                -375.63
No. Observations:                 244   AIC:                             755.3
Df Residuals:                     242   BIC:                             762.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.3307      0.168      7.907      0.0

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [4]:
# Calculando o logaritmo de net_bill
tips['log_net_bill'] = np.log(tips['net_bill'])

# Ajustando o modelo de Regressão de tip_pct por log_net_bill
regression_model_tip_pct = smf.ols(formula='tip_pct ~ log_net_bill', data=tips).fit()

print(regression_model_tip_pct.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Sat, 22 Jul 2023   Prob (F-statistic):           1.62e-13
Time:                        20:24:03   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.6181      0.054     11.444   

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [6]:
# Criando colunas com termos polinomiais de 2º em net_bill
tips['net_bill_squared'] = np.power(tips['net_bill'], 2)

regression_model_tip_pct_poly = smf.ols(formula='tip_pct ~net_bill +  net_bill_squared', data=tips).fit()

print(regression_model_tip_pct_poly.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Sat, 22 Jul 2023   Prob (F-statistic):           1.06e-09
Time:                        20:35:27   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.4851      0.047  

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [7]:
# Criando coluna com logaritmo do tip_pct
tips['log_tip_pct'] = np.log(tips['tip_pct'])

# Ajustando modelo de regressão do log_tip_pct por net_bill
regression_model_log_tip_pct = smf.ols(formula='log_tip_pct ~net_bill', data=tips).fit()

# Calculando o R² do modelo em tip_pct
r_squared_tip_pct = regression_model_log_tip_pct.rsquared

# PRINT
print(f"R² do modelo em tip_pct: {r_squared_tip_pct:.4f}")

R² do modelo em tip_pct: 0.2353


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [9]:
# Ajustando modelo de regressão do log_tip_pct por log_net_bill
regression_model_log_tip_pct_log_net_bill = smf.ols(formula='np.log(tip_pct) ~np.log(net_bill)', data=tips).fit()

# Calculando R² do modelo em tip_pct
r_squared_tip_pct_log_net_bill = regression_model_log_tip_pct_log_net_bill.rsquared

# PRINT
print(f"R² do MOdelo em tip_pct: {r_squared_tip_pct_log_net_bill:.4f}")

R² do MOdelo em tip_pct: 0.2935


###  6. qual modelo resultou no melhor ajuste?

O modelo original teve o maior valor de R² (0.333) sendo o segundo melhor o modelo log do [tip_pct] em função do log de [net_bill] com (0.2935)